File: src/main/java/us/illyohs/civilmagiks/client/render/tile/RenderManaSpring.java
package us.illyohs.civilmagiks.client.render.tile;

import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.VertexBuffer;
import net.minecraft.client.renderer.tileentity.TileEntitySpecialRenderer;
import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
import net.minecraft.util.ResourceLocation;
import org.lwjgl.opengl.GL11;
import us.illyohs.civilmagiks.client.core.util.AssetUtils;
import us.illyohs.civilmagiks.client.core.util.FxUtil;
import us.illyohs.civilmagiks.common.tile.node.TileManaSpring;

import java.awt.*;

public class RenderManaSpring extends TileEntitySpecialRenderer<TileManaSpring> {

    private ResourceLocation TEXTURE = new ResourceLocation("civilmagicks", "textures/models/manaspring.png");

    @Override
    public void renderTileEntityAt(TileManaSpring te, double x, double y, double z, float tick, int destroyStage) {
        renderCircle(te.getManaType().getColor(), x, y, z, tick);
    }

    private void renderCircle(int color, double x, double y, double z, float tick) {

        Tessellator tess = Tessellator.getInstance();
        VertexBuffer buff = tess.getBuffer();

        GlStateManager.pushMatrix();
//        GL11.glStencilFunc();
        bindTexture(TEXTURE);
        GlStateManager.disableCull();
        GlStateManager.enableBlend();
        GlStateManager.blendFunc(770, 1);
        GlStateManager.translate(x+.5, y+.025, z+.5);
        GlStateManager.scale(3,3,3);
        GlStateManager.rotate(90.0f, 1, 0, 0);
        GlStateManager.pushMatrix();

        float spin = tick / 16F;
        GlStateManager.rotate(-spin, 0, 0, 1);
        Color col = new Color(color);
        float r = col.getRed();
        float g = col.getGreen();
        float b = col.getBlue();

        GlStateManager.color(r, g, b, 1f);
        buff.begin(7, DefaultVertexFormats.POSITION_TEX);

        buff.pos(1 / 2f, 1 / 2f, 0.0D).tex(1, 1).endVertex();
        buff.pos(1 / 2f, -1 / 2f, 0.0D).tex(1, 0).endVertex();
        buff.pos(-1 / 2f, -1 / 2f, 0.0D).tex(0, 0).endVertex();
        buff.pos(-1 / 2f, 1 / 2f, 0.0D).tex(0, 1).endVertex();
        tess.draw();

        GlStateManager.popMatrix();

        // GlStateManager.depthMask(true);
        GlStateManager.disableBlend();
        GlStateManager.enableCull();
        GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA);

        GlStateManager.popMatrix();
    }

}

File: src/main/scala/us/illyohs/civilmagiks/client/render/tile/RenderManaSpring.scala
package us.illyohs.civilmagiks.client.render.tile


import net.minecraft.client.renderer.{GlStateManager, Tessellator, VertexBuffer}
import net.minecraft.client.renderer.tileentity.TileEntitySpecialRenderer
import net.minecraft.client.renderer.vertex.DefaultVertexFormats
import net.minecraft.util.ResourceLocation
import org.lwjgl.opengl.GL11
import us.illyohs.civilmagiks.common.tile.TileManaSpring

import scala.swing.Color


object RenderManaSpring extends TileEntitySpecialRenderer[TileManaSpring] {

  private val TEXTURE:ResourceLocation = new ResourceLocation("civilmagicks", "textures/models/manaspring.png")

  override def renderTileEntityAt(te:TileManaSpring, x: Double, y: Double, z: Double, tick: Float, dtick: Int): Unit = {
    val color:Color = new Color(te.getManaType.getColor)

    val tess:Tessellator = Tessellator.getInstance()
    val buff:VertexBuffer = tess.getBuffer

    GlStateManager.pushMatrix()
    bindTexture(TEXTURE)
    GlStateManager.disableCull()
    GlStateManager.enableBlend()
    GlStateManager.disableAlpha()
    GlStateManager.blendFunc(770, 1)
    GlStateManager.color(color.getRed, color.getGreen(), color.getBlue(), 300.0F)
    GlStateManager.translate(x + .5, y + .025, z + .5)
    GlStateManager.scale(10, 10, 10)
    GlStateManager.rotate(90.0f, 1, 0, 0)
    GlStateManager.pushMatrix()

    val spin: Float = tick / 16F
    GlStateManager.rotate(-spin, 0, 0, 1)

    buff.begin(7, DefaultVertexFormats.POSITION_TEX)

    buff.pos(1 / 2f, 1 / 2f, 0.0D).tex(1, 1).endVertex()
    buff.pos(1 / 2f, -1 / 2f, 0.0D).tex(1, 0).endVertex()
    buff.pos(-1 / 2f, -1 / 2f, 0.0D).tex(0, 0).endVertex()
    buff.pos(-1 / 2f, 1 / 2f, 0.0D).tex(0, 1).endVertex()

    tess.draw()

    GlStateManager.popMatrix

    // GlStateManager.depthMask(true);
    GlStateManager.disableBlend
    GlStateManager.enableCull
    GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA)

    GlStateManager.popMatrix
  }
}
